### PR TITLE
[API][Web][SIMS #1338] Reorganize API Controllers - Offering

### PR DESCRIPTION
**Pull request by dheepak-aot**
## Reorganize API Controllers - Offering

image

image

image

- [x] Fixed a scenario(new assessment record in db must be created for only completed applications) missed from #1345 

image

### CODE REVIEW

File: sources/packages/api/src/route-controllers/education-program-offering/education-program-offering.aest.controller.ts [hunk 1]
```diff
@@ -64,7 +134,7 @@ export class EducationProgramOfferingAESTController extends BaseController {
   @Roles(Role.InstitutionApproveDeclineOffering)
   @Patch(":offeringId/assess")
   async assessOffering(
-    @Param("offeringId") offeringId: number,
+    @Param("offeringId", ParseIntPipe) offeringId: number,
```
Comment (ann-aot) [inline sources/packages/api/src/route-controllers/education-program-offering/education-program-offering.aest.controller.ts:137]: 👍

File: sources/packages/api/src/route-controllers/education-program-offering/models/education-program-offering.dto.ts [hunk 2]
```diff
@@ -4,35 +4,135 @@ import {
...
+
+export class StudyBreakInDTO {
+  @Allow()
+  breakStartDate: Date;
+  @Allow()
+  breakEndDate: Date;
+}
 
-export interface SaveOfferingDTO {
+export interface StudyBreaksAndWeeksOutDTO {
```
Comment (ann-aot) [inline sources/packages/api/src/route-controllers/education-program-offering/models/education-program-offering.dto.ts:31]: class?

File: sources/packages/api/src/route-controllers/education-program-offering/models/education-program-offering.dto.ts [hunk 3]
```diff
@@ -65,7 +165,7 @@ export interface ProgramOfferingDto {
   showYearOfStudy?: boolean;
   hasOfferingWILComponent: string;
   offeringWILType?: string;
-  breaksAndWeeks: StudyBreaksAndWeeks;
+  breaksAndWeeks: StudyBreaksAndWeeksOutDTO;
```
Comment (ann-aot) [inline sources/packages/api/src/route-controllers/education-program-offering/models/education-program-offering.dto.ts:168]: pls remove this old DTO interface if its not used

File: sources/packages/api/src/services/education-program-offering/education-program-offering.service.ts [hunk 4]
```diff
@@ -820,19 +802,21 @@ export class EducationProgramOfferingService extends RecordDataModelService<Educ
...
-        application.currentAssessment = {
-          application: { id: application.id } as Application,
-          triggerType: AssessmentTriggerType.OfferingChange,
-          offering: { id: offeringId } as EducationProgramOffering,
-          creator: auditUser,
-          createdAt: currentDate,
-          submittedBy: auditUser,
-          submittedDate: currentDate,
-        } as StudentAssessment;
+        if (application.applicationStatus === ApplicationStatus.completed) {
```
Comment (ann-aot) [inline sources/packages/api/src/services/education-program-offering/education-program-offering.service.ts:805]: how about overwritten application?

File: sources/packages/web/src/services/EducationProgramOfferingService.ts [hunk 5]
```diff
@@ -27,97 +25,105 @@ export class EducationProgramOfferingService {
...
    */
   public async createProgramOffering(
     locationId: number,
     programId: number,
-    data: OfferingDTO,
-  ): Promise<number> {
-    return ApiClient.EducationProgramOffering.createProgramOffering(
+    payload: EducationProgramOfferingAPIInDTO,
+  ): Promise<void> {
+    return ApiClient.EducationProgramOffering.createOffering(
```
Comment (ann-aot) [inline sources/packages/web/src/services/EducationProgramOfferingService.ts:38]: `await ApiClient.EducationProgramOffering.createOffering(
      locationId,
      programId,
      payload,
    ); `?

File: sources/packages/web/src/services/EducationProgramOfferingService.ts [hunk 6]
```diff
@@ -27,97 +25,105 @@ export class EducationProgramOfferingService {
...
+   * @param offeringId offering to be modified.
+   */
   public async updateProgramOffering(
     locationId: number,
     programId: number,
     offeringId: number,
-    data: OfferingDTO,
+    payload: EducationProgramOfferingAPIInDTO,
   ): Promise<void> {
     return ApiClient.EducationProgramOffering.updateProgramOffering(
```
Comment (ann-aot) [inline sources/packages/web/src/services/EducationProgramOfferingService.ts:100]: await?

File: sources/packages/web/src/services/EducationProgramOfferingService.ts [hunk 7]
```diff
@@ -238,7 +185,7 @@ export class EducationProgramOfferingService {
     locationId: number,
     programId: number,
     offeringId: number,
-    payload: OfferingDTO,
+    payload: EducationProgramOfferingAPIInDTO,
   ): Promise<void> {
     return ApiClient.EducationProgramOffering.requestChange(
```
Comment (ann-aot) [inline sources/packages/web/src/services/EducationProgramOfferingService.ts:190]: await?

**Review by ann-aot — COMMENTED** [review]
looks good 👍  added some comments

File: sources/packages/api/src/route-controllers/education-program-offering/models/education-program-offering.dto.ts (hunk 3 above)
Comment (dheepak-aot) [reply sources/packages/api/src/route-controllers/education-program-offering/models/education-program-offering.dto.ts:168]: The old DTO is part offering entity modal to define contract to the JSON column.

File: sources/packages/web/src/services/EducationProgramOfferingService.ts (hunk 7 above)
Comment (henryoforeh-aot) [reply sources/packages/web/src/services/EducationProgramOfferingService.ts:190]: Although not for this PR but I see a return on a void promise.

File: sources/packages/api/src/services/education-program-offering/education-program-offering.service.models.ts [hunk 8]
```diff
@@ -6,7 +6,7 @@ import {
   Application,
 } from "../../database/entities";
 
-export class EducationProgramOfferingModel {
+export class EducationProgramOfferingSummaryModel {
```
[the remark this answers is not in the document]
Comment (dheepak-aot) [reply sources/packages/api/src/services/education-program-offering/education-program-offering.service.models.ts:9]: Not sure. It was already existing. Btw, I changed it to interface.

File: sources/packages/api/src/services/education-program-offering/education-program-offering.service.ts (hunk 4 above)
Comment (dheepak-aot) [reply sources/packages/api/src/services/education-program-offering/education-program-offering.service.ts:805]: Overwritten applications will not be a part of this logic. 
i.e when we fetch applications to submit re-assessment, Cancelled and Overwritten applications are ignored there. 
image

image

File: sources/packages/web/src/services/EducationProgramOfferingService.ts (hunk 7 above)
Comment (dheepak-aot) [reply sources/packages/web/src/services/EducationProgramOfferingService.ts:190]: Yes, technically we can return a void promise. And awaiting for that returned promise will wait until all the execution inside this method is complete.

File: sources/packages/api/src/route-controllers/education-program-offering/models/education-program-offering.dto.ts (hunk 3 above)
Comment (ann-aot) [reply sources/packages/api/src/route-controllers/education-program-offering/models/education-program-offering.dto.ts:168]: if this DTO is used in any controller, please consider renaming it according to our DTO naming convention.

**Review by henryoforeh-aot — APPROVED** [review]

**Review by ann-aot — APPROVED** [review]
Thanks for doing the changes 👍 Good Work @dheepak-aot

File: sources/packages/api/src/route-controllers/education-program-offering/education-program-offering.students.controller.ts [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+   * @returns offerings in client lookup format.
+   */
+  @ApiUnprocessableEntityResponse({
+    description: "Invalid offering intensity.",
+  })
+  @Get(
+    "location/:locationId/education-program/:programId/program-year/:programYearId",
+  )
+  async getProgramOfferingsOptionsList(
+    @Param("locationId") locationId: number,
```
Comment (guru-aot) [inline sources/packages/api/src/route-controllers/education-program-offering/education-program-offering.students.controller.ts:52]: Please add , ParseIntPipe for the Id's

File: sources/packages/api/src/route-controllers/education-program-offering/education-program-offering.students.controller.ts [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+  /**
+   * Get offering start date of a given offering.
+   * @param offeringId offering id
+   * @returns offering with start date value.
+   */
+  @ApiNotFoundResponse({ description: "Offering not found." })
+  @Get(":offeringId")
+  async getProgramOfferingDetails(
+    @Param("offeringId") offeringId: number,
```
Comment (guru-aot) [inline sources/packages/api/src/route-controllers/education-program-offering/education-program-offering.students.controller.ts:76]: Please add , ParseIntPipe for the Id's

File: sources/packages/web/src/services/EducationProgramOfferingService.ts [hunk 11]
```diff
@@ -27,97 +25,105 @@ export class EducationProgramOfferingService {
...
-   * @returns offering summary.
+   * Get summary of offerings for a program and location.
+   * Pagination, sort and search are available on results.
+   * @param locationId offering location.
+   * @param programId offering program.
+   * @param paginationOptions pagination options.
+   * @returns offering summary results.
    */
-  public async getAllEducationProgramOffering(
+  public async getOfferingsSummary(
```
Comment (guru-aot) [inline sources/packages/web/src/services/EducationProgramOfferingService.ts:53]: Please remove public from all the methods

**Pull request merged**